File: crates/volc-desktop/src/view/format.rs
use chrono::{DateTime, Utc};

/// Formats a Unix millisecond timestamp for display.
pub fn timestamp(timestamp_ms: i64) -> String {
    DateTime::<Utc>::from_timestamp_millis(timestamp_ms).map_or_else(
        || "未知时间".to_owned(),
        |value| value.format("%Y-%m-%d %H:%M:%S UTC").to_string(),
    )
}

/// Formats a signed number of seconds as a reset countdown.
pub fn countdown(seconds: i64) -> String {
    if seconds <= 0 {
        return "即将重置".to_owned();
    }
    let days = seconds / 86_400;
    let hours = (seconds % 86_400) / 3_600;
    let minutes = (seconds % 3_600) / 60;
    if days > 0 {
        format!("{days} 天 {hours} 小时")
    } else if hours > 0 {
        format!("{hours} 小时 {minutes} 分")
    } else {
        format!("{minutes} 分 {} 秒", seconds % 60)
    }
}

/// Short countdown for compact displays (e.g. `04:54` or `1天`).
pub fn countdown_short(seconds: i64) -> String {
    if seconds <= 0 {
        return "即将重置".to_owned();
    }
    let days = seconds / 86_400;
    let hours = (seconds % 86_400) / 3_600;
    let minutes = (seconds % 3_600) / 60;
    let secs = seconds % 60;
    if days > 0 {
        format!("{days}天 {hours}时")
    } else if hours > 0 {
        format!("{hours:02}:{minutes:02}")
    } else {
        format!("{minutes:02}:{secs:02}")
    }
}

/// Human-friendly "time ago" relative to `now_ms`.
pub fn relative(fetched_at_ms: i64, now_ms: i64) -> String {
    let delta = now_ms.saturating_sub(fetched_at_ms).max(0) / 1_000;
    if delta < 5 {
        return "刚刚".to_owned();
    }
    if delta < 60 {
        return format!("{delta} 秒前");
    }
    let minutes = delta / 60;
    if minutes < 60 {
        return format!("{minutes} 分钟前");
    }
    let hours = minutes / 60;
    if hours < 24 {
        return format!("{hours} 小时前");
    }
    let days = hours / 24;
    format!("{days} 天前")
}

/// Formats a number with thousands separators and one decimal place.
pub fn number(value: f64) -> String {
    let rounded = (value * 10.0).round() / 10.0;
    let mut s = format!("{rounded:.1}");
    let (int_part, frac) = s.split_once('.').unwrap_or((&s, ""));
    let int = int_part
        .chars()
        .rev()
        .collect::<Vec<_>>()
        .chunks(3)
        .map(|c| c.iter().collect::<String>())
        .collect::<Vec<_>>()
        .join(",");
    let int = int.chars().rev().collect::<String>();
    s = if frac.is_empty() {
        int
    } else {
        format!("{int}.{frac}")
    };
    s
}

/// Formats a percentage with one decimal, tolerating values over 100.
pub fn percent(value: f64) -> String {
    format!("{:.1}%", value)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn countdown_handles_boundaries() {
        assert_eq!(countdown(-1), "即将重置");
        assert_eq!(countdown(59), "0 分 59 秒");
        assert_eq!(countdown(3_661), "1 小时 1 分");
        assert_eq!(countdown(90_000), "1 天 1 小时");
    }

    #[test]
    fn invalid_timestamp_is_safe() {
        assert_eq!(timestamp(i64::MAX), "未知时间");
    }

    #[test]
    fn number_uses_thousands_separators() {
        assert_eq!(number(75780.9), "75,780.9");
        assert_eq!(number(100000.0), "100,000.0");
        assert_eq!(number(24219.1), "24,219.1");
        assert_eq!(number(0.0), "0.0");
    }

    #[test]
    fn relative_descends_gracefully() {
        assert_eq!(relative(0, 0), "刚刚");
        assert_eq!(relative(0, 3_000), "刚刚");
        assert_eq!(relative(0, 30_000), "30 秒前");
        assert_eq!(relative(0, 120_000), "2 分钟前");
    }
}

File: src/lib/format.ts
// Display formatting ported from the archived Iced `view/format.rs`.

/** Formats a Unix millisecond timestamp for display. */
export function timestamp(timestampMs: number): string {
  const date = new Date(timestampMs);
  if (Number.isNaN(date.getTime())) return "未知时间";
  const pad = (value: number) => String(value).padStart(2, "0");
  return (
    `${date.getUTCFullYear()}-${pad(date.getUTCMonth() + 1)}-${pad(date.getUTCDate())} ` +
    `${pad(date.getUTCHours())}:${pad(date.getUTCMinutes())}:${pad(date.getUTCSeconds())} UTC`
  );
}

/** Formats a signed number of seconds as a reset countdown. */
export function countdown(seconds: number): string {
  if (seconds <= 0) return "即将重置";
  const days = Math.floor(seconds / 86_400);
  const hours = Math.floor((seconds % 86_400) / 3_600);
  const minutes = Math.floor((seconds % 3_600) / 60);
  if (days > 0) return `${days} 天 ${hours} 小时`;
  if (hours > 0) return `${hours} 小时 ${minutes} 分`;
  return `${minutes} 分 ${seconds % 60} 秒`;
}

/** Short countdown for compact displays (e.g. `04:54` or `1天`). */
export function countdownShort(seconds: number): string {
  if (seconds <= 0) return "即将重置";
  const days = Math.floor(seconds / 86_400);
  const hours = Math.floor((seconds % 86_400) / 3_600);
  const minutes = Math.floor((seconds % 3_600) / 60);
  const secs = seconds % 60;
  if (days > 0) return `${days}天 ${hours}时`;
  if (hours > 0) return `${String(hours).padStart(2, "0")}:${String(minutes).padStart(2, "0")}`;
  return `${String(minutes).padStart(2, "0")}:${String(secs).padStart(2, "0")}`;
}

/** Human-friendly "time ago" relative to `nowMs`. */
export function relative(fetchedAtMs: number, nowMs: number): string {
  const delta = Math.max(0, Math.floor((nowMs - fetchedAtMs) / 1_000));
  if (delta < 5) return "刚刚";
  if (delta < 60) return `${delta} 秒前`;
  const minutes = Math.floor(delta / 60);
  if (minutes < 60) return `${minutes} 分钟前`;
  const hours = Math.floor(minutes / 60);
  if (hours < 24) return `${hours} 小时前`;
  return `${Math.floor(hours / 24)} 天前`;
}

/** Formats a number with thousands separators and one decimal place. */
export function number(value: number): string {
  return value.toLocaleString("en-US", {
    minimumFractionDigits: 1,
    maximumFractionDigits: 1,
  });
}

/** Formats a percentage with one decimal, tolerating values over 100. */
export function percent(value: number): string {
  return `${value.toFixed(1)}%`;
}

/** Relative "last checked" label from an epoch-milliseconds timestamp. */
export function lastCheckedText(timestampMs: number | null): string {
  if (timestampMs === null) return "从未";
  const seconds = Math.max(0, Math.floor((Date.now() - timestampMs) / 1_000));
  if (seconds <= 59) return "刚刚";
  if (seconds < 3_600) return `${Math.floor(seconds / 60)} 分钟前`;
  if (seconds < 86_400) return `${Math.floor(seconds / 3_600)} 小时前`;
  return `${Math.floor(seconds / 86_400)} 天前`;
}

/** Bytes → human readable size. */
export function bytes(value: number): string {
  if (value < 1024) return `${value} B`;
  const units = ["KB", "MB", "GB"];
  let amount = value;
  let unit = "B";
  for (const next of units) {
    if (amount < 1024) break;
    amount /= 1024;
    unit = next;
  }
  return `${amount.toFixed(1)} ${unit}`;
}
